src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker_Generators.cs

```diff
@@ -0,0 +1,382 @@
+            using var _2 = PooledDictionary<DocumentId, int>.GetInstance(out var documentIdToIndex);
+
+            var infos = infosOpt.Value;
+            foreach (var (documentIdentity, contentIdentity) in infos)
+            {
+                var documentId = documentIdentity.DocumentId;
+                Contract.ThrowIfFalse(documentId.IsSourceGenerated);
+
+                var existingDocument = generatorInfo.Documents.GetState(documentId);
+                if (existingDocument?.Identity == documentIdentity)
+                {
+                    // ensure that the doc we have matches the content expected.
+                    if (existingDocument.GetTextChecksum() == contentIdentity.Checksum &&
+                        existingDocument.SourceText.Encoding?.WebName == contentIdentity.EncodingName &&
+                        existingDocument.SourceText.ChecksumAlgorithm == contentIdentity.ChecksumAlgorithm)
+                    {
+                        continue;
+                    }
+                }
+
+                // Couldn't find a matching generated doc.  Add this to the list to pull down.
+                documentIdToIndex.Add(documentId, documentsToAddOrUpdate.Count);
+                documentsToAddOrUpdate.Add(documentId);
+            }
+
+            // If we produced just as many documents as before, and none of them required any changes, then we can
+            // reuse the prior compilation.
+            if (infos.Length == generatorInfo.Documents.Count &&
+                documentsToAddOrUpdate.Count == 0 &&
+                compilationWithStaleGeneratedTrees != null &&
+                generatorInfo.Documents.States.All(kvp => kvp.Value.ParseOptions.Equals(this.ProjectState.ParseOptions)))
+            {
+                return (compilationWithStaleGeneratedTrees, generatorInfo.WithDocumentsAreFinal(true));
+            }
+
+            // Either we generated a different number of files, and/or we had contents of files that changed. Ensure
+            // we know the contents of any new/changed files.
+            var generatedSourcesOpt = await connection.TryInvokeAsync(
+                solution,
+                this.ProjectState.Id,
+                (service, solutionChecksum, cancellationToken) => service.GetContentsAsync(
+                    solutionChecksum, projectId, documentsToAddOrUpdate.ToImmutable(), cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!generatedSourcesOpt.HasValue)
+                return null;
+
+            var generatedSources = generatedSourcesOpt.Value;
+            Contract.ThrowIfTrue(generatedSources.Length != documentsToAddOrUpdate.Count);
+
+            // Now go through and produce the new document states, using what we have already if it is unchanged, or
+            // what we have retrieved for anything new/changed.
+            using var generatedDocumentsBuilder = TemporaryArray<SourceGeneratedDocumentState>.Empty;
```

Comment:
Perf wise this approach here will re-create the TextDocumentStates every time even if only a single document changed, I'm not sure if it's possible to instead have this instead add/replace the existing ones (and deal with removal) so that way there's a bit less recreation of those dictionaries. But if this was a hotspot you'd see it easily enough in traces.